### PR TITLE
[DATA-277] Update date/time format to match datamanager

### PR DESCRIPTION
**Pull request by JohnN193**
Update slam repo to use the time format that other services are using. Currently setup to still work with our previous time format. 

rdk PR: https://github.com/viamrobotics/rdk/pull/1504

https://viam.atlassian.net/browse/DATA-277

### CODE REVIEW

File: slam-libraries/viam-orb-slam3/orbslam_server_v1.cc [hunk 1]
```diff
@@ -778,13 +779,24 @@ double ReadTimeFromFilename(string filename) {
...
+        return myTime;
+    } else {
+        return (double)thisTime;
+    }
+}
 
-    double myTime = (double)thisTime + sub_sec;
-    return myTime;
+// Checks if using new or old time format
+void FindTimeFormat(string filename) {
```
Comment (JohnN193) [inline slam-libraries/viam-orb-slam3/orbslam_server_v1.cc:795]: function intended to check which format is being used. Can be removed after PR in rdk goes thru

File: slam-libraries/viam-orb-slam3/orbslam_server_v1.h [hunk 2]
```diff
@@ -90,6 +90,9 @@ void ParseAndValidateArguments(const vector<string> &args,
 // Converts UTC time string to a double value.
 double ReadTimeFromFilename(const string filename);
 
+// Checks if using new or old time format
+void FindTimeFormat(string filename);
```
Comment (JohnN193) [inline slam-libraries/viam-orb-slam3/orbslam_server_v1.h:94]: removed after rdk passes

File: slam-libraries/viam-orb-slam3/orbslam_server_v1_main.cc [hunk 3]
```diff
@@ -100,6 +101,8 @@ int main(int argc, char **argv) {
             return 1;
         }
     }
+    viam::utils::FindTimeFormat(
```
Comment (JohnN193) [inline slam-libraries/viam-orb-slam3/orbslam_server_v1_main.cc:104]: remove after rdk passes

File: slam-libraries/viam-cartographer/src/io/read_PCD_file.cc [hunk 4]
```diff
@@ -38,6 +38,12 @@ ReadFile::timedPointCloudDataFromPCDBuilder(std::string file_path,
 
     // KAT NOTE: The file name format for the pcd files is assumed to be, e.g.:
     // rplidar_data_2022-02-05T01_00_20.9874.pcd
+
+    // LEAVING COMMENTED UNTIL READY TO TEST
+    // double time_delta = ReadTimeFromFilename(
```
Comment (JohnN193) [inline slam-libraries/viam-cartographer/src/io/read_PCD_file.cc:43]: Leaving this commented as I did not get the chance to test this fully with cartographer and am unsure if other spots  would need this

File: slam-libraries/viam-orb-slam3/orbslam_server_v1_main.cc [hunk 5]
```diff
@@ -159,6 +162,10 @@ int main(int argc, char **argv) {
         slamService.StartSaveAtlasAsOsa(SLAM.get());
         slamService.ProcessDataOnline(SLAM.get());
         slamService.StopSaveAtlasAsOsa();
+        while (viam::b_continue_session) {
+            this_thread::sleep_for(chrono::microseconds(
+                viam::checkForShutdownIntervalMicroseconds));
+        }
```
Comment (kkufieta) [inline slam-libraries/viam-orb-slam3/orbslam_server_v1_main.cc:168]: [q] Why did you add this? I would think that in the case of `ProcessDataOnline` finishing, the user already tried to kill the program and `b_continue_session` would be `false` anyways?

File: slam-libraries/viam-cartographer/src/io/read_PCD_file.h [hunk 6]
```diff
@@ -13,14 +13,19 @@
 #include "cartographer/sensor/timed_point_cloud_data.h"
 
 namespace viam {
+static const int filenamePrefixLength = 6;
 namespace io {
 
 class ReadFile {
    public:
+    std::string time_format = "%Y-%m-%dT%H:%M:%SZ";
```
Comment (kkufieta) [inline slam-libraries/viam-cartographer/src/io/read_PCD_file.h:21]: [q] Why not have this inside the scope of `ReadTimeFromFilename` since that's where it's used?

File: slam-libraries/viam-cartographer/src/io/read_PCD_file.h [hunk 7]
```diff
@@ -13,14 +13,19 @@
 #include "cartographer/sensor/timed_point_cloud_data.h"
 
 namespace viam {
+static const int filenamePrefixLength = 6;
```
Comment (kkufieta) [inline slam-libraries/viam-cartographer/src/io/read_PCD_file.h:16]: Can you move this into the `viam::io::` namespace?

File: slam-libraries/viam-orb-slam3/orbslam_server_v1.cc [hunk 8]
```diff
@@ -778,13 +779,24 @@ double ReadTimeFromFilename(string filename) {
...
     time_t thisTime = std::mktime(&dt);
+    auto sub_sec_index = filename.find(".");
+    if ((sub_sec_index != string::npos)) {
+        double sub_sec = (double)std::stof(filename.substr(sub_sec_index), &sz);
+        double myTime = (double)thisTime + sub_sec;
+        return myTime;
+    } else {
+        return (double)thisTime;
+    }
+}
```
Comment (kkufieta) [inline slam-libraries/viam-orb-slam3/orbslam_server_v1.cc:792]: Would it be possible to put this also into a `ReadTimeFromFilename` function, together with the `string time_format` definition above?

File: slam-libraries/viam-orb-slam3/orbslam_server_v1.cc [hunk 9]
```diff
@@ -866,16 +878,18 @@ int FindFrameIndex(const std::vector<std::string> &filesRGB,
...
+                  std::gmtime(&t));
     // Save the current atlas map in *.osa style
-    string path_save_file_name =
-        path_to_dir + "/" + camera_name + "_data_" + timestamp + ".0000.osa";
-    return path_save_file_name;
+    auto find_underscores = time_format.find("_");
+    if (find_underscores != string::npos) {
+        return path_to_dir + "/" + camera_name + "_data_" + timestamp +
+               ".0000.osa";
+    }
```
Comment (kkufieta) [inline slam-libraries/viam-orb-slam3/orbslam_server_v1.cc:891]: [q] will this be removed later, too?

**Review by kkufieta — COMMENTED** [review]
Nice work - I have a few questions

File: slam-libraries/viam-orb-slam3/orbslam_server_v1_main.cc (hunk 5 above)
Comment (JohnN193) [reply slam-libraries/viam-orb-slam3/orbslam_server_v1_main.cc:168]: leftovers from some debugging that I forgot about, ty for catching

File: slam-libraries/viam-orb-slam3/orbslam_server_v1.cc (hunk 9 above)
Comment (JohnN193) [reply slam-libraries/viam-orb-slam3/orbslam_server_v1.cc:891]: yeah if we want to remove all support for the previous timestamp then this will be removed as well

File: slam-libraries/viam-orb-slam3/orbslam_server_v1.cc (hunk 8 above)
Comment (JohnN193) [reply slam-libraries/viam-orb-slam3/orbslam_server_v1.cc:792]: these are currently inside the ReadTimeFromFilename function

File: slam-libraries/viam-cartographer/src/io/read_PCD_file.h (hunk 6 above)
Comment (JohnN193) [reply slam-libraries/viam-cartographer/src/io/read_PCD_file.h:21]: because other methods will probably need to know the time format, such as saving the pbstream. Should move this out of `ReadfFile` and into `io`?

File: slam-libraries/viam-orb-slam3/orbslam_server_v1.cc (hunk 8 above)
Comment (JohnN193) [reply slam-libraries/viam-orb-slam3/orbslam_server_v1.cc:792]: as for `string time_format` I do not want to move that into the function, since multiple functions depend on it

File: slam-libraries/viam-cartographer/src/io/read_PCD_file.cc [hunk 10]
```diff
@@ -38,6 +38,12 @@ ReadFile::timedPointCloudDataFromPCDBuilder(std::string file_path,
...
+    //     myYAML.substr(initial_filename.find("_data_") +
```
Comment (tessavitabile) [inline slam-libraries/viam-cartographer/src/io/read_PCD_file.cc:44]: `myYAML` should be `initial_filename`.

File: slam-libraries/viam-cartographer/src/io/read_PCD_file.h (hunk 6 above)
Comment (kkufieta) [reply slam-libraries/viam-cartographer/src/io/read_PCD_file.h:21]: I think for now I'd add it to the function. If it needs to be pulled out in the future for other functions, it can happen then.

File: slam-libraries/viam-orb-slam3/orbslam_server_v1.cc [hunk 11]
```diff
@@ -866,16 +878,18 @@ int FindFrameIndex(const std::vector<std::string> &filesRGB,
...
+    return path_to_dir + "/" + camera_name + "_data_" + timestamp + "Z.osa";
```
Comment (tessavitabile) [inline slam-libraries/viam-orb-slam3/orbslam_server_v1.cc:892]: Is the "Z" not already included in the timestamp?

File: slam-libraries/viam-orb-slam3/orbslam_server_v1_test.cc [hunk 12]
```diff
@@ -133,7 +133,7 @@ BOOST_AUTO_TEST_CASE(ReadTimeFromFilename) {
 }
 
 BOOST_AUTO_TEST_CASE(FindFrameIndex_Closest_no_files) {
-    const string configTimeString = "2022-01-01T01_00_00.0000";
+    const string configTimeString = "2022-01-01T01:00:00.0000";
```
Comment (tessavitabile) [inline slam-libraries/viam-orb-slam3/orbslam_server_v1_test.cc:136]: Why doesn't this use Z?

File: slam-libraries/viam-orb-slam3/orbslam_server_v1.cc [hunk 13]
```diff
@@ -778,13 +779,24 @@ double ReadTimeFromFilename(string filename) {
 
     // Now we read from buffer using get_time manipulator
     // and formatting the input appropriately.
-    ss >> std::get_time(&dt, "%Y-%m-%dT%H_%M_%SZ");
-    double sub_sec =
-        (double)std::stof(filename.substr(filename.find(".")), &sz);
+    ss >> std::get_time(&dt, time_format.c_str());
     time_t thisTime = std::mktime(&dt);
+    auto sub_sec_index = filename.find(".");
+    if ((sub_sec_index != string::npos)) {
```
Comment (tessavitabile) [inline slam-libraries/viam-orb-slam3/orbslam_server_v1.cc:785]: Why will there not always be sub-seconds?

**Review by tessavitabile — COMMENTED** [review]
Nice! A couple comments/questions.

File: slam-libraries/viam-orb-slam3/orbslam_server_v1.cc (hunk 13 above)
Comment (JohnN193) [reply slam-libraries/viam-orb-slam3/orbslam_server_v1.cc:785]: because the .osa files would save with .0000 before, the new time format in rdk would end up removing the sub second portion when loading from an apriori map. I can add a comment in the code to clarify this

File: slam-libraries/viam-orb-slam3/orbslam_server_v1_test.cc (hunk 12 above)
Comment (JohnN193) [reply slam-libraries/viam-orb-slam3/orbslam_server_v1_test.cc:136]: good catch, forgot to include the Z when updating the tests

File: slam-libraries/viam-orb-slam3/orbslam_server_v1.cc (hunk 11 above)
Comment (JohnN193) [reply slam-libraries/viam-orb-slam3/orbslam_server_v1.cc:892]: good catch, forgot to fix this here

File: slam-libraries/viam-orb-slam3/orbslam_server_v1.cc (hunk 13 above)
Comment (tessavitabile) [reply slam-libraries/viam-orb-slam3/orbslam_server_v1.cc:785]: Does that mean this can go away? (Actually, do you think you could put TODOs over everything that will go away?)

Comment (JohnN193) [reply slam-libraries/viam-orb-slam3/orbslam_server_v1.cc:785]: no this one would need to stay, unless we wanted to remove sub seconds from the yaml/osa files entirely

Comment (tessavitabile) [reply slam-libraries/viam-orb-slam3/orbslam_server_v1.cc:785]: Ah, thanks

**Comment by JohnN193** [comment]
Do we want to remove support for the old format/datasets after integration tests in rdk pass? Just so stuff like the hackweek5 dataset still works if we plan to keep using it

File: slam-libraries/viam-orb-slam3/orbslam_server_v1.cc (hunk 13 above)
Comment (tessavitabile) [inline slam-libraries/viam-orb-slam3/orbslam_server_v1.cc:785]: [nit] remove extra parentheses

**Review by tessavitabile — APPROVED** [review]
LGTM mod nit. I don't think we need to continue to support the old format. If we want to, we can convert hackweek5 to the new format.

File: slam-libraries/viam-cartographer/src/io/read_PCD_file.cc [hunk 14]
```diff
@@ -86,7 +86,9 @@ int ReadFile::removeFile(std::string file_path) {
 
 // Converts UTC time string to a double value.
 double ReadFile::ReadTimeFromFilename(std::string filename) {
-    std::string time_format = "%Y-%m-%dT%H:%M:%SZ";
+    // TODO: change time format "%Y-%m-%dT%H:%M:%SZ"
+    // https://viam.atlassian.net/browse/DATA-638
+    std::string time_format = "%Y-%m-%dT%H_%M_%S";
```
Comment (kkufieta) [inline slam-libraries/viam-cartographer/src/io/read_PCD_file.cc:91]: Change this back to the original:

```
std::string time_format = "%Y-%m-%dT%H:%M:%SZ";
```

**Review by kkufieta — APPROVED** [review]
LGTM mod changing back to original `time_format` in `read_PCD_file.cc`
